File: src/types/benchmark.ts
import { z } from "zod";
import { AgentKindSchema } from "@/types/agent";

const BenchmarkFileSchema = z.object({
	/** Portable path within one case. */
	path: z.string(),
	/** App-owned immutable resource. */
	assetId: z.string(),
});

const BenchmarkAssetPreviewSchema = z.object({
	/** Opaque managed identifier requested by the caller. */
	assetId: z.string().min(1),
	/** Complete file size before preview truncation. */
	sizeBytes: z.number().int().nonnegative(),
	/** UTF-8 prefix, or null for binary content. */
	text: z.string().nullable(),
	/** Whether trailing bytes were omitted. */
	truncated: z.boolean(),
});

const BenchmarkCheckKindSchema = z.literal([
	"answer",
	"file_exists",
	"file_text",
	"file_json",
	"python",
]);

const BenchmarkImportPreviewSchema = z.object({
	/** Proposed catalog name. */
	name: z.string(),
	/** Proposed catalog description. */
	description: z.string(),
	/** Optional external attribution. */
	source: z.string().nullable(),
	/** Bounded Case summaries in template order. */
	cases: z.array(
		z.object({
			/** Proposed Case name. */
			name: z.string(),
			/** Per-Agent execution deadline. */
			timeoutMinutes: z.number().int(),
			/** Number of public files copied into the Case workspace. */
			inputFileCount: z.number().int().nonnegative(),
			/** Stable kinds of checks configured for the Case. */
			checkKinds: z.array(BenchmarkCheckKindSchema),
		}),
	),
	/** Total public inputs and private verifier files. */
	fileCount: z.number().int().nonnegative(),
	/** Field-addressable problems retained for draft repair. */
	issues: z.array(
		z.object({
			/** Field path that can be mapped back to the editor. */
			field: z.string(),
			/** Stable import-validation rule identifier. */
			code: z.string(),
		}),
	),
});

const BenchmarkCheckSchema = z.discriminatedUnion("kind", [
	z.object({
		/** Exact final answer check. */
		kind: z.literal("answer"),
		/** Expected answer. */
		expected: z.string(),
	}),
	z.object({
		/** Required output file. */
		kind: z.literal("file_exists"),
		/** Relative output path. */
		path: z.string(),
	}),
	z.object({
		/** Text content check. */
		kind: z.literal("file_text"),
		/** Relative output path. */
		path: z.string(),
		/** Required text. */
		expected: z.string(),
	}),
	z.object({
		/** JSON output check. */
		kind: z.literal("file_json"),
		/** Relative output path. */
		path: z.string(),
		/** Expected JSON text. */
		expected: z.string(),
	}),
	z.object({
		/** Controlled Python validator. */
		kind: z.literal("python"),
		/** Private script asset. */
		script: BenchmarkFileSchema,
	}),
]);

const BenchmarkCaseSchema = z.object({
	/** Unique name in this suite. */
	name: z.string(),
	/** Complete task requirements. */
	prompt: z.string(),
	/** Per-agent time limit. */
	timeoutMinutes: z.number().int(),
	/** Public starting files. */
	inputFiles: z.array(BenchmarkFileSchema),
	/** Automatic scoring criteria. */
	checks: z.array(BenchmarkCheckSchema),
});

const BenchmarkDocumentSchema = z.object({
	/** Portable document format. */
	schemaVersion: z.number().int(),
	/** Display name. */
	name: z.string(),
	/** Purpose of the suite. */
	description: z.string(),
	/** Exactly one selected local tag on publication. */
	tagId: z.string().nullable(),
	/** Optional provenance. */
	source: z.string().nullable(),
	/** Complete ordered suite. */
	cases: z.array(BenchmarkCaseSchema),
});

const BenchmarkTagSchema = z.object({
	/** Stable classification ID. */
	id: z.string(),
	/** Display label. */
	name: z.string(),
	/** Gravity icon export. */
	icon: z.string(),
	/** Protected fallback tag. */
	isSystem: z.boolean(),
});

const BenchmarkSummarySchema = z.object({
	/** Definition route ID. */
	id: z.string(),
	/** Display title. */
	name: z.string(),
	/** Card description. */
	description: z.string(),
	/** Single tag relation. */
	tagId: z.string(),
	/** Backend-owned author. */
	author: z.enum(["platform", "myself"]),
	/** Optional attribution. */
	source: z.string().nullable(),
	/** Whether new mounts are forbidden. */
	archived: z.boolean(),
	/** Latest published version ID. */
	versionId: z.string(),
	/** Latest version number. */
	versionNumber: z.number().int(),
	/** Published case count. */
	caseCount: z.number().int(),
	/** Creation timestamp. */
	createdAtMs: z.number().int(),
});

const BenchmarkDetailSchema = z.object({
	/** Current display metadata. */
	summary: BenchmarkSummarySchema,
	/** Selected immutable version. */
	versionId: z.string(),
	/** Selected version number. */
	versionNumber: z.number().int(),
	/** All selected cases and resources. */
	document: BenchmarkDocumentSchema,
});

const BenchmarkDraftSchema = z.object({
	/** Editor route ID. */
	id: z.string(),
	/** Original definition for an edit. */
	benchmarkId: z.string().nullable(),
	/** Optimistic save revision. */
	revision: z.number().int(),
	/** Incomplete editor content. */
	document: BenchmarkDocumentSchema,
	/** Last successful save. */
	updatedAtMs: z.number().int(),
});

const BenchmarkMountSchema = z.object({
	/** Relationship route ID. */
	id: z.string(),
	/** Owning workspace. */
	workspaceId: z.string(),
	/** Published definition. */
	benchmarkId: z.string(),
	/** Pinned version, never implicitly upgraded. */
	versionId: z.string(),
	/** Mount timestamp. */
	createdAtMs: z.number().int(),
});

const BenchmarkPreviewSchema = z.object({
	/** Checked version. */
	versionId: z.string(),
	/** Display version. */
	versionNumber: z.number().int(),
	/** Suite name. */
	name: z.string(),
	/** Complete case range. */
	cases: z.array(
		z.object({
			/** Stable ordering. */
			position: z.number().int(),
			/** Case label. */
			name: z.string(),
			/** Per-execution limit. */
			timeoutMinutes: z.number().int(),
		}),
	),
	/** Ordered products. */
	agentKinds: z.array(AgentKindSchema),
	/** Case × Agent total. */
	executionCount: z.number().int(),
	/** Explicit file policy. */
	fileAccess: z.enum(["read_only", "allow_edits"]),
	/** Explicit command policy. */
	commandExecution: z.enum(["deny", "ask", "allow"]),
	/** Safe localized blockers. */
	issues: z.array(
		z.object({
			/** Missing prerequisite. */
			code: z.enum([
				"agent_not_installed",
				"agent_not_authenticated",
				"agent_check_failed",
				"asset_unavailable",
				"verifier_unavailable",
			]),
			/** Affected product. */
			agentKind: AgentKindSchema.nullable(),
			/** Affected case position. */
			casePosition: z.number().int().nullable(),
		}),
	),
});

const BenchmarkValidationIssueSchema = z.object({
	/** Field path such as cases.0.prompt. */
	field: z.string(),
	/** Stable publication rule identifier. */
	code: z.string(),
});

const BenchmarkValidationErrorSchema = z.object({
	/** Stable IPC error code for field-addressable publication failures. */
	code: z.literal("BENCHMARK_VALIDATION_FAILED"),
	/** Safe structured error payload consumed by the editor. */
	details: z.object({
		/** Discriminant separating these issues from other IPC details. */
		kind: z.literal("benchmarkValidation"),
		/** Complete bounded publication issues. */
		issues: z.array(BenchmarkValidationIssueSchema),
	}),
});

const BenchmarkTagUsageSchema = z.object({
	/** Definitions reassigned when the selected Tag is deleted. */
	benchmarkCount: z.number().int().nonnegative(),
});

const BenchmarkEvaluationReportSchema = z.object({
	/** Overall result; true only when every public check passed. */
	passed: z.boolean(),
	/** Bounded public checks returned by built-in or controlled validators. */
	checks: z.array(
		z.object({
			/** Stable check kind used by the result UI. */
			kind: z.string().min(1),
			/** Relative artifact path addressed by the check, when applicable. */
			path: z.string().nullable(),
			/** Whether this individual check succeeded. */
			passed: z.boolean(),
			/** Stable safe result message identifier. */
			message: z.string().min(1),
		}),
	),
});

const BenchmarkArtifactFileSchema = z.object({
	/** Portable path within the final execution workspace. */
	path: z.string().min(1),
	/** Final size, or baseline size for a deleted file. */
	sizeBytes: z.number().int().nonnegative(),
	/** Exact comparison against the immutable Case baseline. */
	change: z.enum(["added", "modified", "deleted", "unchanged"]),
});

const BenchmarkArtifactPreviewSchema = z.object({
	/** Portable path within the final execution workspace. */
	path: z.string().min(1),
	/** Complete file size before preview truncation. */
	sizeBytes: z.number().int().nonnegative(),
	/** UTF-8 prefix, or null for binary content. */
	text: z.string().nullable(),
	/** Whether trailing bytes were omitted. */
	truncated: z.boolean(),
});

const BenchmarkArtifactFilesSchema = z.array(BenchmarkArtifactFileSchema);

const BenchmarkExecutionMetricsSchema = z.object({
	/** Full Agent execution duration in milliseconds. */
	totalDurationMs: z.number().int().nonnegative(),
	/** Delay until the first non-empty Agent response delta. */
	timeToFirstTokenMs: z.number().int().nonnegative().nullable(),
	/** Token categories when the source Agent reports them. */
	tokenUsage: z
		.object({
			/** Total input and output tokens. */
			totalTokens: z.number().int().nonnegative(),
			/** Tokens included in model input. */
			inputTokens: z.number().int().nonnegative(),
			/** Input tokens served from cache. */
			cachedInputTokens: z.number().int().nonnegative(),
			/** Input tokens written into cache. */
			cacheWriteInputTokens: z.number().int().nonnegative(),
			/** Tokens included in model output. */
			outputTokens: z.number().int().nonnegative(),
			/** Output tokens consumed by reasoning when reported. */
			reasoningOutputTokens: z.number().int().nonnegative().nullable(),
		})
		.nullable(),
	/** Number of recorded tool calls. */
	toolCallCount: z.number().int().nonnegative(),
	/** Tool calls retained in source start order. */
	toolCalls: z.array(
		z.object({
			/** Stable tool name supplied by the source protocol. */
			name: z.string(),
			/** Wall-clock execution duration in milliseconds. */
			durationMs: z.number().int().nonnegative(),
		}),
	),
});

const BenchmarkTaskDetailSchema = z.object({
	/** Common Task identity and lifecycle. */
	task: z.object({
		/** Stable local Task identifier. */
		id: z.string().min(1),
		/** Workspace that owns the Benchmark Task. */
		workspaceId: z.string().min(1).nullable(),
		/** User-visible Task title. */
		title: z.string().min(1),
		/** Discriminant used by the shared task route. */
		kind: z.literal("benchmark"),
		/** Aggregate persisted lifecycle state. */
		status: z.literal([
			"preparing",
			"running",
			"waiting",
			"completed",
			"failed",
			"stopped",
		]),
		/** Time after which execution configuration cannot change. */
		configurationLockedAtMs: z.number().int().nullable(),
		/** Optional pin time used by the shared sidebar. */
		pinnedAtMs: z.number().int().nullable(),
		/** Creation time in Unix milliseconds. */
		createdAtMs: z.number().int(),
		/** Latest lifecycle update time in Unix milliseconds. */
		updatedAtMs: z.number().int(),
	}),
	/** Published definition used by the Task. */
	benchmarkId: z.string().min(1),
	/** Display name captured for the result header. */
	benchmarkName: z.string().min(1),
	/** Immutable published version executed by the Task. */
	versionId: z.string().min(1),
	/** User-visible version number. */
	versionNumber: z.number().int().positive(),
	/** Historical Task that initiated this Rerun, when present. */
	rerunOfTaskId: z.string().min(1).nullable(),
	/** Whether every planned cell has a complete normal result. */
	resultCompleteness: z.enum(["complete", "incomplete"]),
	/** Stable reason for an incomplete result matrix. */
	completionReason: z.string().nullable(),
	/** Whether the user requested cancellation. */
	cancelRequested: z.boolean(),
	/** Frozen file access policy. */
	fileAccess: z.enum(["read_only", "allow_edits"]),
	/** Frozen command execution policy. */
	commandExecution: z.enum(["deny", "ask", "allow"]),
	/** Aggregate matrix completion counts. */
	progress: z.object({
		/** Number of planned matrix cells. */
		total: z.number().int().nonnegative(),
		/** Number of terminal matrix cells. */
		finished: z.number().int().nonnegative(),
		/** Number of passing evaluations. */
		passed: z.number().int().nonnegative(),
		/** Number of failing evaluations. */
		failed: z.number().int().nonnegative(),
		/** Number of terminal cells without an evaluation verdict. */
		errors: z.number().int().nonnegative(),
	}),
	/** Matrix columns and per-Agent aggregates. */
	agents: z.array(
		z.object({
			/** Stable Task Agent row identifier. */
			id: z.string().min(1),
			/** Local Agent product selected for this column. */
			agentKind: AgentKindSchema,
			/** Stable matrix column order. */
			position: z.number().int().nonnegative(),
			/** Number of passing evaluations for this Agent. */
			passed: z.number().int().nonnegative(),
			/** Number of failing evaluations for this Agent. */
			failed: z.number().int().nonnegative(),
			/** Number of planned Cases for this Agent. */
			total: z.number().int().nonnegative(),
			/** Complete pass rate, absent until every Case is countable. */
			passRate: z.number().min(0).max(1).nullable(),
			/** Sum of reported execution durations. */
			totalDurationMs: z.number().nonnegative(),
			/** Number of executions contributing duration data. */
			durationCoverage: z.number().int().nonnegative(),
			/** Sum of reported token totals. */
			totalTokens: z.number().nonnegative(),
			/** Number of executions contributing token data. */
			tokenCoverage: z.number().int().nonnegative(),
			/** Sum of reported tool-call counts. */
			toolCallCount: z.number().int().nonnegative(),
		}),
	),
	/** Frozen matrix rows in published order. */
	cases: z.array(
		z.object({
			/** Stable Task Case row identifier. */
			id: z.string().min(1),
			/** Published immutable Case identifier. */
			caseId: z.string().min(1),
			/** Stable matrix row order. */
			position: z.number().int().nonnegative(),
			/** User-visible Case name. */
			name: z.string().min(1),
			/** Complete requirements sent to every selected Agent. */
			prompt: z.string(),
			/** Per-execution deadline in minutes. */
			timeoutMinutes: z.number().int().positive(),
		}),
	),
	/** Complete Case × Agent result matrix. */
	executions: z.array(
		z.object({
			/** Stable execution cell identifier. */
			id: z.string().min(1),
			/** Frozen Task Case row used by this cell. */
			taskCaseId: z.string().min(1),
			/** Frozen Task Agent row used by this cell. */
			taskAgentId: z.string().min(1),
			/** Database-constrained lifecycle phase for this execution cell. */
			phase: z.literal([
				"queued",
				"preparing",
				"running",
				"waiting_permission",
				"collecting",
				"evaluating",
				"stopping",
				"finished",
			]),
			/** Verdict, terminal reason, or current phase displayed in the matrix. */
			result: z.string().min(1),
			/** Terminal reason when the cell has no verdict. */
			terminationReason: z.string().nullable(),
			/** Final Agent response when one was produced. */
			responseText: z.string().nullable(),
			/** Typed normalized metrics when the Agent reported them. */
			metrics: BenchmarkExecutionMetricsSchema.nullable(),
			/** Execution start time in Unix milliseconds. */
			startedAtMs: z.number().int().nullable(),
			/** Execution completion time in Unix milliseconds. */
			finishedAtMs: z.number().int().nullable(),
			/** Passed or failed evaluation verdict. */
			verdict: z.enum(["passed", "failed"]).nullable(),
			/** Typed deterministic evaluation report. */
			report: BenchmarkEvaluationReportSchema.nullable(),
		}),
	),
});

const BenchmarkSummariesSchema = z.array(BenchmarkSummarySchema);
const BenchmarkMountsSchema = z.array(BenchmarkMountSchema);
const BenchmarkTagsSchema = z.array(BenchmarkTagSchema);
const BenchmarkDraftIdsSchema = z.array(z.string());
const EmptyBenchmarkResponseSchema = z.null();

type BenchmarkDocument = z.infer<typeof BenchmarkDocumentSchema>;

type BenchmarkFile = z.infer<typeof BenchmarkFileSchema>;

type BenchmarkCheck = z.infer<typeof BenchmarkCheckSchema>;

type BenchmarkAssetPreview = z.infer<typeof BenchmarkAssetPreviewSchema>;

type BenchmarkImportPreview = z.infer<typeof BenchmarkImportPreviewSchema>;

type BenchmarkDetail = z.infer<typeof BenchmarkDetailSchema>;

type BenchmarkSummary = z.infer<typeof BenchmarkSummarySchema>;

type BenchmarkDraft = z.infer<typeof BenchmarkDraftSchema>;

type BenchmarkMount = z.infer<typeof BenchmarkMountSchema>;

type BenchmarkTag = z.infer<typeof BenchmarkTagSchema>;

type BenchmarkPreview = z.infer<typeof BenchmarkPreviewSchema>;

type BenchmarkTaskDetail = z.infer<typeof BenchmarkTaskDetailSchema>;

type BenchmarkArtifactFile = z.infer<typeof BenchmarkArtifactFileSchema>;

type BenchmarkArtifactPreview = z.infer<typeof BenchmarkArtifactPreviewSchema>;

type BenchmarkValidationIssue = z.infer<typeof BenchmarkValidationIssueSchema>;

type BenchmarkTagUsage = z.infer<typeof BenchmarkTagUsageSchema>;

type BenchmarkFilters = {
	/** Literal name/description query. */
	search: string;
	/** OR filters from the tag picker. */
	tagIds: string[];
	/** Only local or platform authors exist. */
	author: "platform" | "myself" | null;
	/** Ordering applies before pagination. */
	sort: "newest" | "updated" | "oldest" | "alphabetical";
};

type BenchmarkPreviewInput = Pick<
	BenchmarkPreview,
	"agentKinds" | "fileAccess" | "commandExecution"
> & {
	/** Owning workspace. */
	workspaceId: string;
	/** Selected mount. */
	mountId: string;
	/** Version shown before the check. */
	expectedVersionId: string;
};

type StartBenchmarkTaskInput = BenchmarkPreviewInput & {
	/** Client retry identity for this exact launch request. */
	idempotencyKey: string;
};

type RerunBenchmarkTaskInput = Pick<
	BenchmarkTaskDetail,
	"fileAccess" | "commandExecution"
> & {
	/** Terminal Benchmark Task whose immutable version is reused. */
	sourceTaskId: string;
	/** Local Agent products selected for the fresh execution matrix. */
	agentKinds: BenchmarkPreview["agentKinds"];
	/** Explicit approval to recreate an absent historical mount. */
	restoreMount: boolean;
	/** Retry identity for this exact submission. */
	idempotencyKey: string;
};

export type {
	BenchmarkDocument,
	BenchmarkFile,
	BenchmarkCheck,
	BenchmarkAssetPreview,
	BenchmarkImportPreview,
	BenchmarkDetail,
	BenchmarkSummary,
	BenchmarkDraft,
	BenchmarkMount,
	BenchmarkTag,
	BenchmarkPreview,
	BenchmarkFilters,
	BenchmarkPreviewInput,
	BenchmarkTaskDetail,
	BenchmarkArtifactFile,
	BenchmarkArtifactPreview,
	BenchmarkValidationIssue,
	BenchmarkTagUsage,
	RerunBenchmarkTaskInput,
	StartBenchmarkTaskInput,
};

export {
	BenchmarkFileSchema,
	BenchmarkAssetPreviewSchema,
	BenchmarkImportPreviewSchema,
	BenchmarkSummarySchema,
	BenchmarkDetailSchema,
	BenchmarkDraftSchema,
	BenchmarkMountSchema,
	BenchmarkTagSchema,
	BenchmarkPreviewSchema,
	BenchmarkTaskDetailSchema,
	BenchmarkArtifactFilesSchema,
	BenchmarkArtifactPreviewSchema,
	BenchmarkValidationErrorSchema,
	BenchmarkTagUsageSchema,
	BenchmarkSummariesSchema,
	BenchmarkMountsSchema,
	BenchmarkTagsSchema,
	BenchmarkDraftIdsSchema,
	EmptyBenchmarkResponseSchema,
};

File: src-tauri/src/domain/benchmark.rs
use serde::{Deserialize, Serialize};
use std::collections::HashSet;

/// Versioned editor and import document; no machine paths or Agent credentials belong here.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub(crate) struct BenchmarkDocument {
    /// Template format understood by this application.
    pub(crate) schema_version: u32,
    /// Catalog title, allowed to be empty in a draft.
    pub(crate) name: String,
    /// Purpose and expected Agent work.
    pub(crate) description: String,
    /// Local classification selected before publication.
    pub(crate) tag_id: Option<String>,
    /// Attribution retained by imports and duplicates.
    pub(crate) source: Option<String>,
    /// Independent cases in display order.
    pub(crate) cases: Vec<BenchmarkCase>,
}

/// A case always starts a new Agent session with its own input files.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub(crate) struct BenchmarkCase {
    /// Unique title within the published suite.
    pub(crate) name: String,
    /// Complete request sent to every participating Agent.
    pub(crate) prompt: String,
    /// Runtime deadline shared by all participants.
    pub(crate) timeout_minutes: u32,
    /// Public inputs; verifier material is stored separately.
    pub(crate) input_files: Vec<BenchmarkFile>,
    /// All checks must pass for this case to pass.
    pub(crate) checks: Vec<BenchmarkCheck>,
}

/// Reference to an immutable application-owned file, independent of the import location.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub(crate) struct BenchmarkFile {
    /// Portable relative path used inside the case workspace.
    pub(crate) path: String,
    /// Opaque identifier assigned when the file was copied into managed storage.
    pub(crate) asset_id: String,
}

/// Saved grading criteria; the Agent cannot supply or replace these rules.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "snake_case", deny_unknown_fields)]
pub(crate) enum BenchmarkCheck {
    Answer {
        /// Expected final answer, compared after trimming edge whitespace.
        expected: String,
    },
    FileExists {
        /// Required output path relative to the case workspace.
        path: String,
    },
    FileText {
        /// Output file to compare exactly.
        path: String,
        /// Expected UTF-8 file content.
        expected: String,
    },
    FileJson {
        /// Output JSON file to compare structurally.
        path: String,
        /// Serialized JSON; parsed only at the verification boundary.
        expected: String,
    },
    Python {
        /// Managed verifier script, never copied into public inputs.
        script: BenchmarkFile,
    },
}

/// Field-addressable validation failures used by the editor.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct BenchmarkValidationIssue {
    /// Field path such as cases.0.name.
    pub(crate) field: String,
    /// Stable reason translated by the frontend.
    pub(crate) code: &'static str,
}

impl BenchmarkDocument {
    /// Applies publication requirements without preventing incomplete draft editing.
    pub(crate) fn publication_issues(&self) -> Vec<BenchmarkValidationIssue> {
        let mut issues = Vec::new();
        if self.schema_version != 1 {
            issue(&mut issues, "schemaVersion", "unsupported_schema");
        }
        if self.name.trim().is_empty() || self.name.chars().count() > 80 {
            issue(&mut issues, "name", "invalid_name");
        }
        if self.description.trim().is_empty() || self.description.chars().count() > 1000 {
            issue(&mut issues, "description", "invalid_description");
        }
        if self
            .tag_id
            .as_deref()
            .is_none_or(|id| id.is_empty() || id == "uncategorized")
        {
            issue(&mut issues, "tagId", "tag_required");
        }
        if self.cases.is_empty() || self.cases.len() > 100 {
            issue(&mut issues, "cases", "invalid_case_count");
        }
        let mut names = HashSet::new();
        for (index, case) in self.cases.iter().enumerate() {
            let field = format!("cases.{index}");
            if case.name.trim().is_empty() || case.name.chars().count() > 120 {
                issue(&mut issues, &format!("{field}.name"), "invalid_name");
            }
            if !names.insert(case.name.trim()) {
                issue(&mut issues, &format!("{field}.name"), "duplicate_case");
            }
            if case.prompt.trim().is_empty() || case.prompt.len() > 16000 {
                issue(&mut issues, &format!("{field}.prompt"), "invalid_prompt");
            }
            if !(1..=60).contains(&case.timeout_minutes) {
                issue(
                    &mut issues,
                    &format!("{field}.timeoutMinutes"),
                    "invalid_timeout",
                );
            }
            if case.checks.is_empty() || case.checks.len() > 32 {
                issue(&mut issues, &format!("{field}.checks"), "checks_required");
            }
            if case.input_files.len() > 256 {
                issue(
                    &mut issues,
                    &format!("{field}.inputFiles"),
                    "too_many_files",
                );
            }
            let mut paths = HashSet::new();
            for file in &case.input_files {
                if !safe_relative_path(&file.path) || !safe_asset_id(&file.asset_id) {
                    issue(&mut issues, &format!("{field}.inputFiles"), "unsafe_path");
                }
                if !paths.insert(file.path.to_lowercase()) {
                    issue(
                        &mut issues,
                        &format!("{field}.inputFiles"),
                        "duplicate_path",
                    );
                }
            }
            for check in &case.checks {
                match check {
                    BenchmarkCheck::Answer { expected }
                        if expected.trim().is_empty() || expected.len() > 65536 =>
                    {
                        issue(&mut issues, &format!("{field}.checks"), "invalid_expected")
                    }
                    BenchmarkCheck::FileExists { path }
                    | BenchmarkCheck::FileText { path, .. }
                    | BenchmarkCheck::FileJson { path, .. }
                        if !safe_relative_path(path) =>
                    {
                        issue(&mut issues, &format!("{field}.checks"), "unsafe_path")
                    }
                    BenchmarkCheck::FileJson { expected, .. }
                        if expected.len() > 65536
                            || serde_json::from_str::<serde::de::IgnoredAny>(expected).is_err() =>
                    {
                        issue(&mut issues, &format!("{field}.checks"), "invalid_json")
                    }
                    BenchmarkCheck::FileText { expected, .. } if expected.len() > 65536 => {
                        issue(&mut issues, &format!("{field}.checks"), "invalid_expected")
                    }
                    BenchmarkCheck::Python { script }
                        if !safe_relative_path(&script.path)
                            || !safe_asset_id(&script.asset_id) =>
                    {
                        issue(&mut issues, &format!("{field}.checks"), "unsafe_path")
                    }
                    _ => {}
                }
            }
        }
        issues
    }
}

/// Portable file names reject Windows drive syntax and traversal on every host.
pub(crate) fn safe_relative_path(path: &str) -> bool {
    !path.is_empty()
        && path.len() <= 1024
        && !path.contains(['\\', ':', '\0'])
        && path.split('/').all(|part| {
            !part.is_empty() && part != "." && part != ".." && !part.ends_with([' ', '.'])
        })
}

/// Asset identifiers never act as arbitrary paths into the user's computer.
pub(crate) fn safe_asset_id(id: &str) -> bool {
    !id.is_empty()
        && id.len() <= 100
        && id
            .bytes()
            .all(|byte| byte.is_ascii_alphanumeric() || byte == b'-')
}

/// Keeps validation reports compact and independent of UI language.
fn issue(issues: &mut Vec<BenchmarkValidationIssue>, field: &str, code: &'static str) {
    issues.push(BenchmarkValidationIssue {
        field: field.to_string(),
        code,
    });
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn publication_rejects_missing_criteria_duplicate_cases_and_escaping_paths() {
        let case = BenchmarkCase {
            name: "Same".into(),
            prompt: "Do the work".into(),
            timeout_minutes: 10,
            input_files: vec![BenchmarkFile {
                path: "../secret".into(),
                asset_id: "asset-1".into(),
            }],
            checks: vec![],
        };
        let document = BenchmarkDocument {
            schema_version: 1,
            name: "Suite".into(),
            description: "Measure work".into(),
            tag_id: Some("tag-1".into()),
            source: None,
            cases: vec![case.clone(), case],
        };
        let issues = document.publication_issues();
        assert!(issues.iter().any(|issue| issue.code == "unsafe_path"));
        assert!(issues.iter().any(|issue| issue.code == "duplicate_case"));
        assert!(issues.iter().any(|issue| issue.code == "checks_required"));
    }
    #[test]
    fn publication_rejects_invalid_json_criteria() {
        let document = BenchmarkDocument {
            schema_version: 1,
            name: "JSON suite".to_string(),
            description: "Validate output".to_string(),
            tag_id: Some("tag-1".to_string()),
            source: None,
            cases: vec![BenchmarkCase {
                name: "Report".to_string(),
                prompt: "Create report.json".to_string(),
                timeout_minutes: 10,
                input_files: Vec::new(),
                checks: vec![BenchmarkCheck::FileJson {
                    path: "report.json".to_string(),
                    expected: "{".to_string(),
                }],
            }],
        };
        assert!(document
            .publication_issues()
            .iter()
            .any(|issue| issue.code == "invalid_json"));
    }
}

/// Benchmark catalog value exchanged between persistence and application services.
#[derive(Debug, Clone, PartialEq, Eq)]
pub(crate) struct BenchmarkTag {
    /// Stable local classification identifier.
    pub(crate) id: String,
    /// Display name of the classification.
    pub(crate) name: String,
    /// Gravity icon export name.
    pub(crate) icon: String,
    /// Whether this immutable fallback tag is owned by the application.
    pub(crate) is_system: bool,
}

/// Benchmark catalog value exchanged between persistence and application services.
#[derive(Debug, Clone, PartialEq, Eq)]
pub(crate) struct BenchmarkDraft {
    /// Stable editor draft identifier.
    pub(crate) id: String,
    /// Published definition being edited, when present.
    pub(crate) benchmark_id: Option<String>,
    /// Optimistic concurrency revision.
    pub(crate) revision: i64,
    /// Editable definition including incomplete cases.
    pub(crate) document: BenchmarkDocument,
    /// Last save time in UTC milliseconds.
    pub(crate) updated_at_ms: i64,
}

/// Benchmark catalog value exchanged between persistence and application services.
#[derive(Debug, Clone, PartialEq, Eq)]
pub(crate) struct BenchmarkSummary {
    /// Stable definition identifier.
    pub(crate) id: String,
    /// Current catalog title.
    pub(crate) name: String,
    /// Current catalog description.
    pub(crate) description: String,
    /// Exactly one local classification.
    pub(crate) tag_id: String,
    /// Platform or myself, assigned by the application.
    pub(crate) author: String,
    /// Optional source attribution.
    pub(crate) source: Option<String>,
    /// Whether new mounts are disabled.
    pub(crate) archived: bool,
    /// Latest published immutable version.
    pub(crate) version_id: String,
    /// Monotonic content version number.
    pub(crate) version_number: i64,
    /// Number of cases in the latest version.
    pub(crate) case_count: i64,
    /// Definition creation time in UTC milliseconds.
    pub(crate) created_at_ms: i64,
}

/// Benchmark catalog value exchanged between persistence and application services.
#[derive(Debug, Clone, PartialEq, Eq)]
pub(crate) struct BenchmarkDetail {
    /// Current catalog metadata.
    pub(crate) summary: BenchmarkSummary,
    /// Version explicitly selected for this detail.
    pub(crate) version_id: String,
    /// Selected content version number.
    pub(crate) version_number: i64,
    /// Immutable case content with current display metadata.
    pub(crate) document: BenchmarkDocument,
}

/// Benchmark catalog value exchanged between persistence and application services.
#[derive(Debug, Clone, PartialEq, Eq)]
pub(crate) struct BenchmarkMount {
    /// Stable mount identifier.
    pub(crate) id: String,
    /// Workspace owning the mount.
    pub(crate) workspace_id: String,
    /// Mounted benchmark definition.
    pub(crate) benchmark_id: String,
    /// Explicitly pinned content version.
    pub(crate) version_id: String,
    /// Mount creation time in UTC milliseconds.
    pub(crate) created_at_ms: i64,
}
